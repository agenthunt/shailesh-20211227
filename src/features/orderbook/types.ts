export interface NormalizedOrderSideType {
  id: string;
  price: number;
  size: number;
  total: number;
}

export type NormalizedOrderSideEntryType = {
  [key: string]: NormalizedOrderSideType;
};

export enum FeedType {
  BOOK_UI_1 = 'book_ui_1',
  BOOK_UI_1_SNAPSHOT = 'book_ui_1_snapshot',
}

export enum ProductIdType {
  PI_XBTUSD = 'PI_XBTUSD',
  PI_ETHUSD = 'PI_ETHUSD',
}

export interface OrderbookState {
  bids: OrderSideType[];
  asks: OrderSideType[];
  numLevels: number;
  displaySize: number;
  currentProductIds: ProductIdType[];
  subscribeToOrderbookInProgress: boolean;
  subscribeToOrderbookError: any;
  unSubscribeToOrderbookInProgress: boolean;
  unSubscribeToOrderbookError: any;
}

export enum EventType {
  SUBSCRIBE = 'subscribe',
  UNSUBSCRIBE = 'unsubscribe',
  SUBSCRIBED = 'subscribed',
}

export type MessageOrderSideType = [number, number];
export type OrderSideType = [number, number, number];

export interface MessagePayloadType {
  feed: FeedType;
  event?: EventType;
  product_ids?: ProductIdType[];
  bids?: MessageOrderSideType[];
  asks?: MessageOrderSideType[];
  numLevels?: number;
}

export type OrderListItemType = {
  price: string;
  size: string;
  total: string;
  depth: number;
};

export enum OrderListColumnSortOrder {
  TSP = 'TSP',
  PST = 'PST',
}
