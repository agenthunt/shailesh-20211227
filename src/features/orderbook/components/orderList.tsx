import React from 'react';
import {FlatList, StyleSheet, Text, View} from 'react-native';
import {AccentColor} from '../../../theme/colors';
import {
  OrderListColumnSortOrder,
  OrderListItemType,
  OrderSideType,
} from '../types';
import {OrderListItem} from './orderListItem';

const OrderListHeaderRow = ({
  columnOrder,
}: {
  columnOrder: OrderListColumnSortOrder;
}) => {
  const priceColumn = (
    <View style={styles.column}>
      <Text style={styles.headerPrice}>PRICE</Text>
    </View>
  );
  const sizeColumn = (
    <View style={styles.column}>
      <Text style={styles.headerSize}>SIZE</Text>
    </View>
  );

  const totalColumn = (
    <View style={styles.column}>
      <Text style={styles.headerTotal}>TOTAL</Text>
    </View>
  );
  return (
    <View style={styles.headerRowContainer}>
      {columnOrder === OrderListColumnSortOrder.PST ? (
        <>
          {priceColumn}
          {sizeColumn}
          {totalColumn}
        </>
      ) : (
        <>
          {totalColumn}
          {sizeColumn}
          {priceColumn}
        </>
      )}
    </View>
  );
};

export function OrderList({
  items,
  depthColor,
  inverted = false,
  priceColumnColor = '#0e2018',
  columnOrder = OrderListColumnSortOrder.PST,
  hideHeader = false,
}: {
  items: OrderListItemType[];
  depthColor: string;
  inverted?: boolean;
  priceColumnColor?: string;
  columnOrder?: OrderListColumnSortOrder;
  hideHeader?: boolean;
}) {
  const renderOrderListItem = ({item}: {item: OrderListItemType}) => {
    return (
      <OrderListItem
        item={item}
        depthColor={depthColor}
        priceColumnColor={priceColumnColor}
        columnOrder={columnOrder}
      />
    );
  };
  return (
    <View style={styles.container}>
      {!hideHeader ? <OrderListHeaderRow columnOrder={columnOrder} /> : null}
      <FlatList
        data={items}
        renderItem={renderOrderListItem}
        keyExtractor={(item: OrderListItemType) => String(item.price)}
        inverted={inverted}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  itemContainer: {
    flexDirection: 'row',
    padding: 4,
  },
  headerRowContainer: {
    flexDirection: 'row',
    marginBottom: 16,
  },
  column: {
    flex: 0.33,
    alignItems: 'flex-end',
    paddingRight: 20,
  },
  textTotal: {
    color: '#fff',
  },
  textSize: {
    color: '#fff',
  },
  textPrice: {
    color: AccentColor,
  },
  headerTotal: {
    color: '#98a6af',
    borderBottomWidth: 0.3,
    borderStyle: 'dashed',
    borderRadius: 1,
    borderBottomColor: '#98a6af',
    alignSelf: 'flex-end',
  },
  headerSize: {
    color: '#98a6af',
    borderBottomWidth: 0.3,
    borderStyle: 'dashed',
    borderRadius: 1,
    borderBottomColor: '#98a6af',
    alignSelf: 'flex-end',
  },
  headerPrice: {
    color: '#98a6af',
    borderBottomWidth: 0.3,
    borderStyle: 'dashed',
    borderRadius: 1,
    borderBottomColor: '#98a6af',
    alignSelf: 'flex-end',
  },
});
