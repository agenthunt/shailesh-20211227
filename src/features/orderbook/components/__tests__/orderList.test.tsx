/**
 * @format
 */

import 'react-native';
import React from 'react';

import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';

import {OrderList} from '../orderList';
import {OrderListColumnSortOrder, OrderListItemType} from '../../types';

describe('OrderList component testing', () => {
  it('renders default state of OrderList correctly', () => {
    const items: OrderListItemType[] = [
      {
        price: '1000',
        size: '50',
        total: '50',
        depth: 4.3,
      },
      {
        price: '1005',
        size: '100',
        total: '150',
        depth: 13,
      },
      {
        price: '1010',
        size: '1000',
        total: '1150',
        depth: 100,
      },
    ];
    const tree = renderer
      .create(
        <OrderList
          items={items}
          depthColor="#641c25"
          priceColumnColor="#eb4057"
          columnOrder={OrderListColumnSortOrder.PST}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
