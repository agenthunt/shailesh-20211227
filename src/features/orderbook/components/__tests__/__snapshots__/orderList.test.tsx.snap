// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderList component testing renders default state of OrderList correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 0.33,
          "paddingRight": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "flex-end",
            "borderBottomColor": "#98a6af",
            "borderBottomWidth": 0.3,
            "borderRadius": 1,
            "borderStyle": "dashed",
            "color": "#98a6af",
            "fontFamily": "LED SCREEN GTAMBLOG CAPS+ 2.0",
            "fontSize": 12,
            "textAlign": "center",
            "textAlignVertical": "center",
          }
        }
      >
        PRICE
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 0.33,
          "paddingRight": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "flex-end",
            "borderBottomColor": "#98a6af",
            "borderBottomWidth": 0.3,
            "borderRadius": 1,
            "borderStyle": "dashed",
            "color": "#98a6af",
            "fontFamily": "LED SCREEN GTAMBLOG CAPS+ 2.0",
            "fontSize": 12,
            "textAlign": "center",
            "textAlignVertical": "center",
          }
        }
      >
        SIZE
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 0.33,
          "paddingRight": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "flex-end",
            "borderBottomColor": "#98a6af",
            "borderBottomWidth": 0.3,
            "borderRadius": 1,
            "borderStyle": "dashed",
            "color": "#98a6af",
            "fontFamily": "LED SCREEN GTAMBLOG CAPS+ 2.0",
            "fontSize": 12,
            "textAlign": "center",
            "textAlignVertical": "center",
          }
        }
      >
        TOTAL
      </Text>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "depth": 4.3,
          "price": "1000",
          "size": "50",
          "total": "50",
        },
        Object {
          "depth": 13,
          "price": "1005",
          "size": "100",
          "total": "150",
        },
        Object {
          "depth": 100,
          "price": "1010",
          "size": "1000",
          "total": "1150",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    invertStickyHeaders={false}
    inverted={false}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
                "flexDirection": "row",
                "padding": 4,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "height": 28,
                "position": "absolute",
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#641c25",
                  "borderColor": "rgba(255, 255, 255, 0.1)",
                  "borderWidth": 0.1,
                  "height": "100%",
                  "width": "430%",
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "paddingRight": 20,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#29baa5",
                    "fontFamily": "LED SCREEN GTAMBLOG CAPS+ 2.0",
                    "fontSize": 12,
                    "textAlign": "center",
                    "textAlignVertical": "center",
                  },
                  Object {
                    "color": "#eb4057",
                  },
                ]
              }
            >
              1000
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "paddingRight": 20,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#fff",
                  "fontFamily": "LED SCREEN GTAMBLOG CAPS+ 2.0",
                  "fontSize": 12,
                  "textAlign": "center",
                  "textAlignVertical": "center",
                }
              }
            >
              50
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "paddingRight": 20,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#fff",
                  "fontFamily": "LED SCREEN GTAMBLOG CAPS+ 2.0",
                  "fontSize": 12,
                  "textAlign": "center",
                  "textAlignVertical": "center",
                }
              }
            >
              50
            </Text>
          </View>
        </View>
      </View>
      <View
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
                "flexDirection": "row",
                "padding": 4,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "height": 28,
                "position": "absolute",
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#641c25",
                  "borderColor": "rgba(255, 255, 255, 0.1)",
                  "borderWidth": 0.1,
                  "height": "100%",
                  "width": "1300%",
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "paddingRight": 20,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#29baa5",
                    "fontFamily": "LED SCREEN GTAMBLOG CAPS+ 2.0",
                    "fontSize": 12,
                    "textAlign": "center",
                    "textAlignVertical": "center",
                  },
                  Object {
                    "color": "#eb4057",
                  },
                ]
              }
            >
              1005
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "paddingRight": 20,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#fff",
                  "fontFamily": "LED SCREEN GTAMBLOG CAPS+ 2.0",
                  "fontSize": 12,
                  "textAlign": "center",
                  "textAlignVertical": "center",
                }
              }
            >
              100
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "paddingRight": 20,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#fff",
                  "fontFamily": "LED SCREEN GTAMBLOG CAPS+ 2.0",
                  "fontSize": 12,
                  "textAlign": "center",
                  "textAlignVertical": "center",
                }
              }
            >
              150
            </Text>
          </View>
        </View>
      </View>
      <View
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
                "flexDirection": "row",
                "padding": 4,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "height": 28,
                "position": "absolute",
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#641c25",
                  "borderColor": "rgba(255, 255, 255, 0.1)",
                  "borderWidth": 0.1,
                  "height": "100%",
                  "width": "10000%",
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "paddingRight": 20,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#29baa5",
                    "fontFamily": "LED SCREEN GTAMBLOG CAPS+ 2.0",
                    "fontSize": 12,
                    "textAlign": "center",
                    "textAlignVertical": "center",
                  },
                  Object {
                    "color": "#eb4057",
                  },
                ]
              }
            >
              1010
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "paddingRight": 20,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#fff",
                  "fontFamily": "LED SCREEN GTAMBLOG CAPS+ 2.0",
                  "fontSize": 12,
                  "textAlign": "center",
                  "textAlignVertical": "center",
                }
              }
            >
              1000
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "paddingRight": 20,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#fff",
                  "fontFamily": "LED SCREEN GTAMBLOG CAPS+ 2.0",
                  "fontSize": 12,
                  "textAlign": "center",
                  "textAlignVertical": "center",
                }
              }
            >
              1150
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
